### DISCUSSION TITLE
Set environment in script `before_` not picked up when applying dotfiles

### DISCUSSION THREAD

**Discussion by latipun7** [twpayne/chezmoi#2331] [Q&A] [answered]
I'm always trying to bootstrap my password manager (bitwarden) before applying dotfiles. But, it always failed because in my script `before_` I'm doing pulling custom node manager and install that in directory and add the proper `export PATH`  and other environments. Then trying to login to bitwarden which is exporting `BW_SESSION` so user didn't asked to login repeatedly in this session.
This is what always not worked. All those environments I set, not picked up when applying dotfiles, which result in errors, like bitwarden not found in `PATH` or login session not picked up, so I need to enter the password multiple times.

https://github.com/twpayne/chezmoi/blob/097a88c6da1cc5118b4544ae4119e8d2b6fe25f1/assets/chezmoi.io/docs/user-guide/use-scripts-to-perform-actions.md?plain=1#L21-L24
Reading those commits make me want to open this question. How can I bootstrap environment variables in `before_` script, so the next steps, like applying and `after_` script can pick that environment variables?

Currently I'm extracting that script like this:
https://github.com/latipun7/dotfiles/blob/c026cfe5bd369cb6adb1e58f648ab7befe26e9c4/install.sh#L174-L185
but, that is not ideal, so many repeated scripts, I can't leverage using chezmoi `before_` script which can accept templating.

Thank you.

**Comment by twpayne** [comment, marked as answer]
Sadly, this is not generally possible due to the way Linux/UNIX processes work.

Environment variables are passed from parent processes to their children. They are not shared and do not pass from children to their parents. When a child process sets an environment variable (e.g. adding an extra directory to `$PATH` or setting a variable like `$BW_SESSION`) this is only visible to the child process and not the parent [1]. So, the chezmoi process that spawns your `run_` script does not see any changes that the child process makes.

The Bitwarden password manager stores persistent logins in an environment variable, so, by design, a child Bitwarden process cannot login on behalf of a parent process. Some other password managers instead use persistent daemons which are able to communicate with both parent and child processes and so do not have this limitation.

So, your only effective option is to execute the Bitwarden login in a parent process of chezmoi, which sadly cannot be achieved with a `run_before_` script spawned by chezmoi itself. I would open a discussion with the Bitwarden developers about this.

[1] In some cases, it is possible for a parent process to read a child processes's environment variables by reading the contents of `/proc/$CHILD_PID/environ`, which, in theory, would allow a parent process to take environment variable changes from a child process. However, this only works in specific circumstances and is likely fragile.

**Reply by latipun7** [reply]
Yeah, thank you for your clarification. I thought so too, that's why I'm extracting the script and spawning chezmoi from there. Also, thank you if you would bring this up to the Bitwarden developers for a discussion.

[1] `/proc/$CHILD_PID/environ` is new to me, thank you. But like you said, that is something not practical.

**Comment by twpayne** [comment]
Just to clarify, I don't use Bitwarden and so will not raise this with the Bitwarden developers. The discussion should be started by a Bitwarden user.
